### DISCUSSION TITLE
Change the spoiler pattern

### DISCUSSION THREAD

**Discussion by nulldg** [stoatchat/discussions#540] [Feature Suggestions]
**Proposal**: Change the spoiler pattern from `!!text!!` to `||text||`.

Using a common punctuation symbol in a remark pattern is a bad idea because it frequently interferes with natural communication. In this case, it interferes with the posting of multiple exclamation marks. For example, a message such as "Yeah!! Thank you!!" will result in "Thank you" getting spoilered erroneously. This occurs no matter how many exclamation marks are added.

This is why Discord had the foresight to use an non-natural pattern, `||`, instead. Revolt doesn't need to follow this convention and may opt to use something else. Anything will work as long as it uses non-natural characters, but is still cosmetically satisfactory. However, in the interests of reducing the number of people asking how spoilers work in Revolt, I would strongly recommend following the convention.

---

Alternative solutions include:
- A setting which automatically escapes all exclamation mark pairs being sent by the user, thereby effectively disabling the generation of spoilers for that user.
- A setting which enables the use of `||` instead of `!!`. This would automatically convert `||` into `!!` in any message being sent, while escaping `!!` into `\!\!`. This effectively solves the problem for the average user without affecting anything else.
- Make the pattern more robust against potentially accidental spoilering, such as prohibiting the generation of spoiler remarks which have been padded with whitespaces on either or both sides. This will still break some existing spoilers but not as many.

---

Optional accessory suggestions include:
- Progressively update existing spoilers in chatlogs in order to ensure they won't be retroactively broken by the change.
- Users familiar with or who prefer the `!!` pattern may have a legacy option in the settings in order to automatically translate the old style into the proposed one.
- The KaTeX pattern suffers from a similar issue, and could be mitigated by changing its pattern from `$katex$` to `$$katex$$`. The current KaTeX pattern frequently interferes with talk involving figures in dollars, such as prices and finance. This affects all dollars across multiple countries, not just USD. If accepted, this change would best be implemented at the same time as changing the spoiler pattern.

**Comment by MasicoreLord** [comment]
Not sure about the KaTeX though, as changing the delimiters/patterns w/o backwards compatibility similar to the spoilers would mess w/ existing messages

**Reply by nulldg** [reply]
agreed. this is why the issue focuses primarily on spoilers and not katex. it's just a point i've seen brought up in the revolt lounge so i've decided to advocate for it here.

the chatlog correction process should use the same remark pipeline as revite in order to know exactly what to modify and what to leave alone, so unless i'm mistaken, the conversion should be perfect in theory.

another solution to fix backwards compatibility is to make the client-side remark pipeline use the old format for old messages, and the new format for new messages. it's technical debt and it feels gross, but it would work.
